airbyte-cdk/python/airbyte_cdk/sources/declarative/auth/token.py

```diff
@@ -115,3 +118,121 @@ def token(self) -> str:
+    `"Specific Header": "Session Token Value"`
+
+    Attributes:
+        api_url (Union[InterpolatedString, str]): Base api url of source
+        username (Union[InterpolatedString, str]): The username
+        config (Config): The user-provided configuration as specified by the source's spec
+        password (Union[InterpolatedString, str]): The password
+        header (Union[InterpolatedString, str]): Specific header of source for providing session token
+        options (Mapping[str, Any]): Additional runtime parameters to be used for string interpolation
+        session_token (Union[InterpolatedString, str]): Session token generated by user
+        session_token_response_key (Union[InterpolatedString, str]): Key for retrieving session token from api response
+    """
+
+    api_url: Union[InterpolatedString, str]
+    header: Union[InterpolatedString, str]
+    session_token: Union[InterpolatedString, str]
+    session_token_response_key: Union[InterpolatedString, str]
+    username: Union[InterpolatedString, str]
+    config: Config
+    options: InitVar[Mapping[str, Any]]
+    password: Union[InterpolatedString, str] = ""
+
+    def __post_init__(self, options):
+        self._username = InterpolatedString.create(self.username, options=options)
+        self._password = InterpolatedString.create(self.password, options=options)
+        self._api_url = InterpolatedString.create(self.api_url, options=options)
+        self._header = InterpolatedString.create(self.header, options=options)
+        self._session_token = InterpolatedString.create(self.session_token, options=options)
+        self._session_token_response_key = InterpolatedString.create(self.session_token_response_key, options=options)
+
+        self.logger = logging.getLogger("airbyte")
+
+    @property
+    def auth_header(self) -> str:
+        return self._header.eval(self.config)
+
+    @property
+    def token(self) -> str:
+        if self._session_token.eval(self.config):
+            if self.is_valid_session_token():
+                return self._session_token.eval(self.config)
+        if self._password.eval(self.config) and self._username.eval(self.config):
+            username = self._username.eval(self.config)
+            password = self._password.eval(self.config)
+            session_token_response_key = self._session_token_response_key.eval(self.config)
+            api_url = f"{self._api_url.eval(self.config)}session"
+
+            self.logger.info("Using generated session token by username and password")
+            return get_new_session_token(api_url, username, password, session_token_response_key)
+
+        raise ConnectionError("Invalid credentials: session token is not valid or provide username and password")
+
+    def is_valid_session_token(self) -> bool:
+        try:
+            response = requests.get(
+                f"{self._api_url.eval(self.config)}user/current", headers={self.auth_header: self._session_token.eval(self.config)}
```

Comment:
the path will need to be configurable. This could be done by adding a field "validate_session_url", which could be configured as
`"/user/current"` in the case of metabase

Reply:
done